### PR TITLE
Fix b/36685309 Allow prebuilt filters to access params

### PR DESCRIPTION
**Pull request by BrettMichaelJohnson**

### CODE REVIEW

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 1]
```diff
@@ -14,6 +14,8 @@
 
 package com.google.enterprise.adaptor.prebuilt;
 
+import com.google.common.base.Strings;
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Really, stop doing this.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 2]
```diff
@@ -14,6 +14,8 @@
...
+import com.google.common.collect.ImmutableSet;
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Unused, as far as I can tell.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 3]
```diff
@@ -41,82 +44,132 @@
...
+   * be specified in the document's {@code Metadata}, but some key/values of
+   * interest (e.g. ContentType, DocId) exist in the document's {@code params}.
    *
    * <p>Example configuration:
-   * <pre><code>overwrite=false
-   *3.from=colour
-   *3.to=color
-   *5.from=author
-   *5.to=contributors</code></pre>
+   * <pre><code>
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:107]: Why this change?

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 4]
```diff
@@ -41,82 +44,132 @@
...
+    boolean overwrite
+        = Boolean.parseBoolean(getTrimmedValue(config, "overwrite"));
     List<KeyPairing> copies = parseCopies(config);
     if (copies.isEmpty()) {
       log.warning("No entries listed to be copied");
     }
     return new CopyTransform(copies, overwrite, false);
   }
 
+  @Deprecated
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: These deprecated methods all need a javadoc tag saying that copy should be used instead. I'm not convinced it's a necessary change, either. These are MetadataTransforms, after all.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 5]
```diff
@@ -205,26 +287,61 @@ public String toString() {
...
       dest = to;
     }
 
+    @Override
+    public String toString() {
+      return "(from=" + src + ",to=" + dest + ")";
+    }
+  }
+
+  /** Contains a key in keyset. */
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Maybe s/Contains/Describes/? s/in/in a/?

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 6]
```diff
@@ -205,26 +287,61 @@ public String toString() {
...
+    Key(String key, Keyset keyset) {
+      if (key == null || keyset == null) {
+        throw new NullPointerException();
+      }
+      this.key = key;
+      this.keyset = keyset;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:333]: You usually want to override hashCode if you override equals. I know this is a private class (now).

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 7]
```diff
@@ -335,48 +467,81 @@ public static MetadataTransform replaceMetadata(
...
+    }
+
+    private void replaceInParams(String key, Map<String, String> params) {
+      String original = getTrimmedValue(params, key);
+      if (original == null) {
+        log.log(Level.FINE, "No param value for {0}. Skipping", key);
+        return;
       }
+      log.log(Level.FINE,
+          "Replacing param value of {0} that match {1} with {2}: {3}",
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: s/match/matches/, but this whole log message seems deceptive (along with its metadata predecessor). Something like "Replacing {1} with {2} in param value of {0}: {3}", and similarly for the metadata log message.

**Review by wiarlawd — CHANGES_REQUESTED** [review]
Partial review. I haven't gotten to the tests yet.

John L

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 8]
```diff
@@ -41,82 +44,132 @@
...
    * configuration entries (where {@code X} is an integer). The value for each
-   * is a metadata key. Copies are applied in the increasing order of the
-   * integers.
+   * is a metadata or param  key. Copies are applied in the increasing order of
+   * the integers. The config keys {@code "X.from.keyset"} and
+   * {@code "X.to.keyset"} may be set to {@code metadata} or to {@code params}
+   * to restrict the source and destination to only {@code Metadata} or
+   * {@code params}, respectively.  Most keys/values of interest will normally
+   * be specified in the document's {@code Metadata}, but some key/values of
+   * interest (e.g. ContentType, DocId) exist in the document's {@code params}.
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: s/ContentType/Content-Type/

This was copied from SkipDocumentFilter, and might need to be changed in RegexFilter, too.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 9]
```diff
@@ -14,6 +14,8 @@
 
 package com.google.enterprise.adaptor.prebuilt;
 
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableSet;
 import com.google.enterprise.adaptor.Metadata;
 import com.google.enterprise.adaptor.MetadataTransform;
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Collection is also an unused import.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 8 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 2 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 1 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 5 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 6 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:333]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 7 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 3 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:107]: Improved readability and consistency with the JavaDoc in the other prebuilt transforms.  But more importantly, I then forgot to add another example using the params, which I have now done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 4 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: removed short overloads

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 9 above)
Comment (wiarlawd) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: bump

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 3 above)
Comment (wiarlawd) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:107]: Not buying readability in the source for a style guide violation.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 10]
```diff
@@ -174,24 +228,53 @@ private CopyTransform(List<KeyPairing> copies, boolean overwrite,
...
-          log.log(Level.FINER, "Preexisting values for {0}. Combining: {1}",
-              new Object[] {kp.dest, destValues});
-          values.addAll(destValues);
+        switch (kp.dest.keyset) {
+          case METADATA:
+            Set<String> destValues = metadata.getAllValues(kp.dest.key);
+            if (!overwrite && !destValues.isEmpty()) {
+              log.log(Level.FINER, "Preexisting values for {0}. Combining: {1}",
+                  new Object[] {kp.dest, destValues});
+            values.addAll(destValues);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Misindented.

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java [hunk 11]
```diff
@@ -74,18 +74,238 @@ public void testCopy() {
...
+      golden.put("colour", "black");
+      golden.put("color", "black");
+      paramsGolden = Collections.unmodifiableMap(golden);
+    }
+    Metadata metadata = new Metadata();
+    Map<String, String> params = new HashMap<String, String>();
+    params.put("colour", "black");
+    transform.transform(metadata, params);
+    assertEquals(paramsGolden, params);
+  }
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java]: Assert the metadata is still empty?

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 12]
```diff
@@ -174,24 +228,53 @@ private CopyTransform(List<KeyPairing> copies, boolean overwrite,
...
+            metadata.set(kp.dest.key, values);
+            break;
+          case PARAMS:
+            String value = values.iterator().next();
+            if (values.size() > 1) {
+              log.log(Level.FINER,
+                  "Multiple values for {0}. Using first value of {1} for {2}",
+                  new Object[] { kp.src, value, kp.dest });
+            }
+            params.put(kp.dest.key, value);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: You're not accounting for overwrite here.

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java [hunk 13]
```diff
@@ -74,18 +74,238 @@ public void testCopy() {
...
+    config.put("1.to.keyset", "params");
+    config = Collections.unmodifiableMap(config);
+
+    MetadataTransform transform = PrebuiltTransforms.copyMetadata(config);
+
+    final Map<String, String> paramsGolden;
+    {
+      Map<String, String> golden = new HashMap<String, String>();
+      golden.put("colour", "black");
+      golden.put("color", "black");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java:211]: As discussed, without overwrite this should stay red.

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java [hunk 14]
```diff
@@ -74,18 +74,238 @@ public void testCopy() {
...
+    config.put("1.to", "color");
+    config.put("1.to.keyset", "params");
     config = Collections.unmodifiableMap(config);
 
+    MetadataTransform transform = PrebuiltTransforms.copyMetadata(config);
+
+    final Map<String, String> paramsGolden;
+    {
+      Map<String, String> golden = new HashMap<String, String>();
+      golden.put("color", "black");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java:336]: Same here, should stay red without overwrite.

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java [hunk 15]
```diff
@@ -118,6 +338,146 @@ public void testMove() {
   }
 
   @Test
+  public void testMoveFromParamsToMetadata() {
+    Map<String, String> config = new HashMap<String, String>();
+    config.put("overwrite", "true");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java]: Why are all the move tests specifying overwrite=true?

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 16]
```diff
@@ -335,48 +468,76 @@ public static MetadataTransform replaceMetadata(
...
+          new Object[] {key, toMatch, replacement, original});
+      Set<String> values = new HashSet<String>(original);
+      for (String value : original) {
+        String newValue = toMatch.matcher(value).replaceAll(replacement);
+        if (overwrite) {
+          values.remove(value);
         }
-        log.log(Level.FINE, "After replacing: {0}", values);
-        metadata.set(key, values);
+        values.add(newValue);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:528]: I haven't thought deeply about it, but it seems like the set semantics used for values here could get us into trouble with overwrite. For example, suppose the original values are "w" and "ww", and the pattern replacement is "." to "$0$0" (the use of $ seems undocumented, as goes group 0; what am I missing?). Then "w" => "ww", and "w" is erased. Then "ww" => "www" and "ww" is erased. Now only "www" is left, but we started with two values.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 17]
```diff
@@ -335,48 +468,76 @@ public static MetadataTransform replaceMetadata(
...
+      }
+      metadata.set(key, values);
+      log.log(Level.FINE, "After replacing metadata values for {0}: {1}",
+          new Object [] {key, values});
+    }
+
+    private void replaceInParams(String key, Map<String, String> params) {
+      String original = getTrimmedValue(params, key);
```
Comment (wiarlawd) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:536]: I guess overwrite=false should just be a no-op here?

**Review by wiarlawd — CHANGES_REQUESTED** [review]

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 3 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:107]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 10 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: Done.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 12 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: The JavaDoc says "If the destination is a param key, overwrite is ignored and its value is always replaced."

I might be convinced that overwrite=false makes a param final, if that would really be a desirable behavior.  The default value of overwrite is false, so making params write-once might be surprising behavior to the users.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 17 above)
Comment (BrettMichaelJohnson) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:536]: ibid

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java (hunk 11 above)
Comment (BrettMichaelJohnson) [reply test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java]: Done.

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java (hunk 15 above)
Comment (BrettMichaelJohnson) [reply test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java]: Copy/paste.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java [hunk 18]
```diff
@@ -174,24 +228,53 @@ private CopyTransform(List<KeyPairing> copies, boolean overwrite,
     @Override
     public void transform(Metadata metadata, Map<String, String> params) {
       for (KeyPairing kp : copies) {
-        Set<String> values = metadata.getAllValues(kp.src);
+        Set<String> values = new TreeSet<String>();
```
Comment (BrettMichaelJohnson) [inline src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: No longer need this TreeSet, since dropping support for metadata-or-params

**Comment by wiarlawd** [comment]
While I'm thinking of it, s/prebuilt filters/prebuilt transforms/, or perhaps even better, PrebuiltTransforms.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 16 above)
Comment (wiarlawd) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java:528]: Let's have a separate CL to fix this bug, by writing into a separate collection the way copy does.

File: src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java (hunk 12 above)
Comment (wiarlawd) [reply src/com/google/enterprise/adaptor/prebuilt/PrebuiltTransforms.java]: OK, my preference is for the default value of overwrite to be false for metadata, but true for params. I think that produces the least surprising results, but still lets you explicitly set overwrite=false if you want to not change a previously set value.

I was thinking that the constructor argument and field could be a Boolean: TRUE, FALSE, or null, where null is interpreted as FALSE by dest.keyset == METADATA, and TRUE by dest.keyset == PARAMS.

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java [hunk 19]
```diff
@@ -478,13 +546,69 @@ public void testMoveFromParamToParam() {
...
+    config.put("1.to.keyset", "params");
+    config = Collections.unmodifiableMap(config);
+
+    MetadataTransform transform = PrebuiltTransforms.moveMetadata(config);
+
+    final Metadata metadataGolden = new Metadata().unmodifiableView();
+    final Map<String, String> paramsGolden;
+    {
+      Map<String, String> golden = new HashMap<String, String>();
+      golden.put("color", "black");
```
Comment (wiarlawd) [inline test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java:592]: I was surprised that colour was gone in this case, but I think it's just a random choice, and we don't want Yet Another Parameter for this. Should we include a note in the moveTransform metadata that if the destination is a param and overwrite is false, the source is still removed?

**Review by wiarlawd — APPROVED** [review]

File: test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java (hunk 19 above)
Comment (BrettMichaelJohnson) [reply test/com/google/enterprise/adaptor/prebuilt/PrebuiltTransformsTest.java:592]: I was wrestling with this. I *could* try to avoid the delete if the move was a no-op, but overwrite == false for params is just messed up enough that I am not sure it is worth the effort.

**Review by PapaJoltOmega — APPROVED** [review]
Thank you
